File: 2021/day1/lib/day1.ex
defmodule Day1 do
  @moduledoc false

  def parse(path) do
    File.stream!(path)
    |> Enum.map(&String.trim(&1))
    |> Enum.map(&String.to_integer(&1))
  end

  def depth_increment_count(input, step \\ 1) do
    Enum.zip_reduce([input, Enum.drop(input, step)], 0, fn [a, b], acc ->
      if b > a do
        acc + 1
      else
        acc
      end
    end)
  end

  def trios_increment(input) do
    depth_increment_count(input, 3)
  end
end

input = Day1.parse("lib/input.txt")
IO.puts(Day1.depth_increment_count(input))
IO.puts(Day1.trios_increment(input))

File: 2021/day1/julia/day1.jl
function get_input(path)
  Base.readlines(path)
end

function process(input)
  collect(Base.Iterators.map(n -> parse(Int64, n), input))
end 

function get_windows(processed, window_size, step)
  ((@view processed[i:i+window_size-1]) for i in 1:step:length(processed) - window_size + step)
end

function incremented(windows)
  collect(Base.Iterators.filter(n -> first(n) < last(n), windows))
end

single_increments = "../lib/input.txt" |> get_input |> process |> p -> get_windows(p, 2, 1) |> incremented

trio_increments = "../lib/input.txt" |> get_input |> process |> p -> get_windows(p, 4, 1) |> incremented

println(length(single_increments))
println(length(trio_increments))

∑Δ(i) = i |> diff .|> >(0) |> sum |> println

input = parse.(Int, readlines("../lib/input.txt"))

windows = [ sum(input[i:i+2]) for i=1:length(input)-2 ]

return [∑Δ(input), ∑Δ(windows)]
